Enhancement: Hierarchical System Properties

Traderjoe95 opened the pull request:
Fixes #129

vietj commented:
can you document this in the index.adoc file ?

also I think we should change the hierarchical name to rather something that mentions that now qualified names are accepted.

vietj commented on vertx-config/src/test/java/io/vertx/config/impl/spi/SystemPropertiesConfigStoreTest.java:
```diff
@@ -41,6 +41,14 @@ public void init() {
     System.setProperty("true", "true");
     System.setProperty("false", "false");
+
+    System.setProperty("hierarchical.key", "value");
```
where is the result asserted ?

Traderjoe95 replied:
This is asserted by the call to `ConfigChecker.check` on line 95

Traderjoe95 replied:
I also added some more assertions to the flat parsing case (lines 79-88)

Traderjoe95 commented:
> can you document this in the index.adoc file ?

I added the documentation and also added some assertions to the test cases. The test now asserts that the properties are indeed parsed to a flat object when `hierarchical = false`, and that the flat properties are still parsed correctly when `hierarchical = true`

> also I think we should change the hierarchical name to rather something that mentions that now qualified names are accepted.

In my opinion, `hierarchical` explains the purpose of the attrbute pretty well. I'm not sure that something like `qualified` will be much clearer. However, maybe I just didn't come up with a good name yet. If we change this, it should also be changed on the `PropertiesConfigProcessor` to preserve consistent naming.

Traderjoe95 commented on vertx-config/src/main/java/io/vertx/config/spi/utils/JsonObjectHelper.java:
```diff
@@ -113,7 +115,11 @@ private static JsonArray asJsonArray(String s) {
       } catch (Exception e) {
         return null;
       }
+    } else if (!s.startsWith("[") && !s.endsWith("]") && s.contains(",")) {
+      // Allow same comma-separated syntax as for .properties files
+      return asJsonArray("[" + s + "]");
     }
```
I added this case for array parsing to be more consistent with the parsing method used by the `PropertiesConfigProcessor`. However, there are still differences, as the latter does not allow the full JSON array syntax (including the square braces) when `hierarchical = true`.

I really think that the `PropertiesConfigProcessor` should also be using `JsonObjectHelper.convert(String)` to parse values. That would make for easier reasoning about the results. Additionally, it would be an option to drop the custom properties parsing logic from `PropertiesConfigProcessor.HierarchicalPropertiesReader` in favor of letting the JDK do the parsing and convert the resulting `Properties` using `JsonObjectHelper.from(...)`. That would be the most consistent option, in my opinion. What do you think, @vietj?

vietj replied:
I think it's a good idea, can we proceed in this PR ?

Traderjoe95 replied:
Yes, I will implement it as proposed

Traderjoe95 replied:
Done

Traderjoe95 commented:
I see the check is failing, but that doesn't seem to be related with my changes. There's some test failing in the `vertx-config-vault` module.

vietj commented:
ok @Traderjoe95

vietj commented:
I will check locally the tests

vietj commented:
I fixed the spring config cloud server issues in the CI

vietj commented:
the current vault failures are due to https://bugs.java.com/bugdatabase/view_bug.do?bug_id=JDK-8266261

vietj commented:
can you rebase and force push on latest master ? now tests should pass

vietj commented:
ping @Traderjoe95

Traderjoe95 commented:
Hey, sorry it looks like I accidentally destroyed the history when rebasing. I will try to fix it

Traderjoe95 commented:
Okay, the commits are back. Do you want me to clean up the history or is the merge commit fine?

vietj commented:
@Traderjoe95 if you can squash commits and sign the ECA, it would be great

Traderjoe95 commented:
Of course, will do.

Regarding the ECA, I already signed it. The failure is related to one of the commits I pulled in from master

vietj commented on vertx-config-spring-config-server/src/main/java/io/vertx/config/spring/SpringConfigServerStore.java:
```diff
@@ -97,6 +97,7 @@
     if (authHeaderValue != null) {
       options.addHeader("Authorization", authHeaderValue);
     }
+    System.out.println("REQUESTING " + path);
```
remove System.out.println

vietj replied:
well I thin I did introduce that myself

Traderjoe95 commented:
Sure, I will create a new one during the day
